codebase/src/compiler/checker-inference.md

```diff
@@ -0,0 +1,245 @@
+instantiating a generic function with some specific type. For example:
+
+``` ts
+declare function setup<T>(config: { initial(): T }): T
+setup({ initial() { return "last" } })
+```
+
+First checks `{ initial() { return "last" } }` to get `{ initial():
+string }`. By matching `T` in `{ initial(): T }` with `string` in `{
+initial(): string }`, it infers that `T` is `string`, making the
+second line the same as if the author had written:
+
+``` ts
+setup<string>({ initial() { return "last" } })
+```
+
+Meaning that the compiler then checks that
+`{ initial() { return "last" } }` is assignable to
+`{ initial(): string }`.
+
+### Walkthrough
+
+Type parameter inference starts off in `inferTypeArguments`, where
+the first step in type parameter inference is to get the type of all
+the arguments to the function whose parameters are being inferred. In
+the above example, the checker says that the type of
+`{ initial() { return "last" } }` is `{ initial(): string }`. This
+type is called the **source** type, since it is the source of
+inferences. It's matched with the parameter type `{ initial(): T }`.
+This is the **target** type -- it contains type parameters which are
+the target of the process.
+
+Type parameter inference is a pairwise walk of the two types, looking
+for type parameters in the target, matching them to corresponding
+types in the source. The type is walked structurally sort of like a tree
+is elsewhere in the compiler.
+
+1. `inferTypes` gets called on each source/target pair with
+   argument=source/parameter=target. There's only one pair here:
+   `{ initial(): string }` and `{ initial(): T }`.
+2. Since both sides are object types, `inferFromProperties` looks
+   through each property of the target and looks for a match in the
+   source. In this case both have the property `initial`.
+3. `initial`'s type is a signature on both sides
+   (`() => T/() => string`), so inference goes to `inferFromSignature`, which
+   recursively infers from the return type.
+4. Now the source/target pair is `T/string`. Since the source is a
+   lone type parameters, we add `string` to the list of candidates for
+   `T`.
+
+Once all the parameters have had `inferTypes` called on them,
+`getInferredTypes` condenses each candidate array to a single type,
+via `getUnionType` in this case. `T`'s candidates array is `[string]`,
+so `getUnionType` immediately returns `string`.
+
+### Other considerations
+
+#### Method of Combining Candidate Arrays
+
+Only inference to return types, `keyof T` and mapped type constraints
+(which are usually `keyof` too) produce a union. All other locations
+call the custom code `getCommonSupertype`, which more or less does
+what it says. Note that object types are always unioned together
+first, regardless of inference position.
+
+#### Interference Between Contextual Typing and Type Parameter Inference
+
+Type parameter inference actually operates in two passes. The first
```

Comment:
Well... I feel like this is simplified some. There's two passes _per argument_, and sometimes we essentially defer work to the next argument and let the next argument's passes decide if we should lock in inferences (like when we're flowing detached generics through a signature). That's what the `SkipGenericFunctions` inference flag sort-of implies. Inference skips its work, and contextual typing picks up the heavy lifting for the next argument (or results get locked in at the end of signature inference). It's... Weird. Lots of work gets deferred throughout this process.

Reply:
I didn't feel like trying to re-understand the details and then explain it, so I think it's good enough for now. We might come back later and flesh this out.